### PR TITLE
Add federation builder extension and related schema configuration code

### PR DESCRIPTION
**Pull request by Shane32**
Part 5 of #3921 

## Summary

This PR adds a `.AddFederation()` extension method to `IGraphQLBuilder` along with `FederationSettings` class to configure a schema for federation.  Sytnax:

```cs
.AddFederation(string version, Action<LinkConfiguration>? configureLinkDirective = null)
```

The version is required and expects a format like "1.0" or "2.3".  Having this parameter be required prevents changes to a user's schema when updating GraphQL.NET; we can always add support for a new version without causing a breaking change.  Note that the code will accept versions newer than recognized versions, but will not import any definitions for the unrecognized version.

To facilitate this, the extension method calls the following new extension methods and schema node visitors, which are all `internal`:
- `schema.AddFederationLink()` - adds the `@link` directive to the schema
- `schema.AddFederationTypesAndDirectives()` - adds the federation directive definitions to the schema (`@key`, `@shareable`, etc)
- `FederationServiceSchemaNodeVisitor` - adds the `_service` field to the query type
- `FederationEntitiesSchemaNodeVisitor` - configures the `_Entity` type and adds the `_entities` field to the query type

Having these additional marked `internal` reduces the number of breaking changes in the future.  We can of course expose this to the public API upon request.

This PR also marks the prior federation classes as obsolete, and updates the existing federation schema-first samples as follows:
- Sample 1 uses the new `.AddFederation()` method, still creating a Federation v1 subgraph schema
- Sample 2 continues to use the obsolete `FederatedSchemaBuilder` class

In v9 we can remove all of the obsolete federation code.

## Todo

- [x] Add unit tests
- [x] Add migration notes

### CODE REVIEW

**Comment by Shane32** [comment]
@gao-artur I could use your general thoughts on the design of this PR.  See PR summary above.  Thanks!

**Comment by gao-artur** [comment]
> Directive importing
There's no way to configure it so that @key is imported as @fed_key.
How important is this feature?.

I think it's a must for v8. Otherwise, users with conflicting directive names won't be able to use Federation.

**Comment by Shane32** [comment]
> > Directive importing
> > There's no way to configure it so that @key is imported as @fed_key.
> > How important is this feature?.
> 
> I think it's a must for v8. Otherwise, users with conflicting directive names won't be able to use Federation.

Any ideas on how we might design the API to allow for this?

**Comment by gao-artur** [comment]
FederationSettings + Visitor sounds like a good idea. Am I missing something that makes it difficult to implement?

**Comment by Shane32** [comment]
Well, I mean, like how do we design the `FederationSettings` class?

For example:

```cs
settings.Mappings.Add("@key", "@fed_key");
```

**Comment by Shane32** [comment]
I can play around with it, but if you have any ideas, lmk.

**Comment by gao-artur** [comment]
What we are looking for is actually configuring [`@link`](https://specs.apollo.dev/link/v1.0/) directive. How about this?
```c#
var settings = new FederationSettings
{
    SdlPrintOptions = new PrintOptions
    {
        IncludeFederationTypes = false
    },
    Directives = new DirectiveSettings
    {
        Link = new Link("https://specs.apollo.dev/federation/v2.3")
        {
            For = Purpose.EXECUTION,
            As = "fed",
            Import =
            [
                DirectiveImport.Key.As("@uniqueKey"),
                DirectiveImport.Shareable
            ]
        }
    }
};
```

The `DirectiveImport` is defined like this
```c#
public class DirectiveImport
{
    public static DirectiveImport Key { get; } = new("@key");

    public static DirectiveImport Shareable { get; } = new("@shareable");

   // other known directives

    public DirectiveImport(string name, string alias)
        : this(name)
    {
        Alias = alias;
    }

    public DirectiveImport(string name)
    {
        Name = name;
    }

    public string Name { get; }

    public string? Alias { get; }

    public DirectiveImport As(string alias) => new(Name, alias);

    // TODO: support namespace
    public override string ToString() => Alias == null
        ? $"\"{Name}\""
        : $"{{name: \"{Name}\", as: \"{Alias}\" }}";

    public static implicit operator DirectiveImport(string name) => new(name);
}
```

Since the `@link` can point to a non-standard spec, the user can use `DirectiveImport` contractors or implicit operator to configure custom directives
```c#
Import =
[
    new DirectiveImport("xxx", "yyy"),
    "zzz"
]
```

And since the `@link` directive is repeatable, we need to allow multiple `Link` configurations
```c#
public class DirectiveSettings
{
    public Link? Link { get; set; }

    public Link[]? AdditionalLinks { get; set; }
}
```

The `Purpose` is also not limited to `EXECUTION` and `SECURITY`, so it can be defined in a similar way to enumerate known values and allow extension with custom purposes.

**Comment by gao-artur** [comment]
If you say that advenced configuration can be done in another place and not in `AddFederation` then I agree. But there must be a way for advenced configuration without manually adding every possible configs to FederationSattings.

**Comment by Shane32** [comment]
@gao-artur This is ready for review and to be merged, I believe.  After our other recent changes, it simplified the `AddFederation` method very nicely, I'm quite pleased with the design at this point.

After #3975 is merged, I'll need to create a test to ensure that if `@key` is configured with an alias, it is properly detected by `FederationEntitiesSchemaNodeVisitor`.  I can do a separate PR after these are both merged, or if #3975 is merged first then I can do it within this PR.

**Comment by Shane32** [comment]
> @gao-artur This is ready for review and to be merged, I believe. After our other recent changes, it simplified the `AddFederation` method very nicely, I'm quite pleased with the design at this point.
> 
> After #3975 is merged, I'll need to create a test to ensure that if `@key` is configured with an alias, it is properly detected by `FederationEntitiesSchemaNodeVisitor`. I can do a separate PR after these are both merged, or if #3975 is merged first then I can do it within this PR.

I made the change in #3921

**Comment by gao-artur** [comment]
Cool! Will review it soon!

**Comment by Shane32** [comment]
> > @gao-artur This is ready for review and to be merged, I believe. After our other recent changes, it simplified the `AddFederation` method very nicely, I'm quite pleased with the design at this point.
> > After #3975 is merged, I'll need to create a test to ensure that if `@key` is configured with an alias, it is properly detected by `FederationEntitiesSchemaNodeVisitor`. I can do a separate PR after these are both merged, or if #3975 is merged first then I can do it within this PR.
> 
> I made the change in #3921

Updated it here

**Review by gao-artur — COMMENTED** [review]
Are there any validations to ensure user doesn't use directives from a higher version than imported?

File: src/GraphQL.Federation.SchemaFirst.Sample2/Program.cs [hunk 1]
```diff
@@ -47,6 +47,8 @@ private static ISchema BuildSchema(IServiceProvider serviceProvider)
         var reader = new StreamReader(stream);
         var schemaString = reader.ReadToEnd();
 
+        // note: this demonstrates GraphQL.NET v7 and prior configuration methods
+#pragma warning disable CS0618 // Type or member is obsolete
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL.Federation.SchemaFirst.Sample2/Program.cs:51]: yes, done

**Comment by Shane32** [comment]
> Are there any validations to ensure user doesn't use directives from a higher version than imported?

Indirectly, yes.  The directives will not be configured for the schema.  During schema initialization, the applied directives will not be able to be matched to configured directives, and initialization checks will fail.  I can add a test to be sure.

**Comment by gao-artur** [comment]
Amazing! Looks really good!

**Review by gao-artur — APPROVED** [review]

**Pull request merged**